### DISCUSSION TITLE
About further charge at 100% SOC when fitting experiment data

### DISCUSSION THREAD

**Discussion by RuiheLi** [pybamm-team/PyBaMM#2040] [Q&A]
Hi PyBaMM,

Recently I am trying to fit the experiment data and find that error will occur when I try to charge the cell at 100% SOC. The code is is follows:

```
Model_9 = pb.lithium_ion.DFN()
chemistry = pb.parameter_sets.Chen2020
Para_0 = pb.ParameterValues(chemistry=chemistry)
V_max = 4.2;        V_min = 2.5;
Experiment_Breakin= pb.Experiment( [    # at 25 degreeC
    (f"Discharge at 0.2 C until {V_min} V",
    "Rest for 4 hours",
    f"Charge at 0.2 C until {V_max} V" , 
    f"Hold at {V_max} V until C/100",
    "Rest for 2 hours")        ]   *1
    )  
Experiment_RPT_Long    = pb.Experiment( [      # under 25 degreeC
    (
        f"Charge at 0.3 C until {V_max} V", 
        f"Hold at {V_max} V until C/100",
        "Rest for 2 hours",  
        f"Discharge at 0.1 C until {V_min} V",  
        "Rest for 6 hours",  
        f"Charge at 0.1 C until {V_max} V",
        "Rest for 10 minutes",  
    ) ] * 1 )
Sol_NonDry=[];  Model_NonDry=[];
Sim_0 = pb.Simulation(Model_9, experiment = Experiment_Breakin,parameter_values = Para_0,solver = pb.CasadiSolver(),)       
Sol_0 = Sim_0.solve();
Sol_NonDry.append(Sol_0); Model_NonDry.append(Model_9);

Model_1 = Model_9.set_initial_conditions_from(Sol_0, inplace=False)
Sim_1 = pb.Simulation(Model_1, experiment = Experiment_RPT_Long,parameter_values = Para_0,solver = pb.CasadiSolver(),)   
Sol_1 = Sim_1.solve(); 
Sol_NonDry.append(Sol_1); Model_NonDry.append(Model_1);
```

I understand it is a bit wired to do this in modelling. But that is quite normal in experiments. I know that if I delete the first two steps in Experiment_RPT_Long (further charge), the error will disappear. But that will make it more complicated to transfer the real test procedure into modelling steps. I wonder whether there are any easy way to solve this? Big thanks in advance!

**Comment by hamidbdr** [comment]
I think that the cell voltage is already at Vmax so when you try to recharge to Vmax it breaks. If you want to continue charging the Cell, you should define cutt off voltage(on the second experiment) greater than your current Vmax

**Reply by RuiheLi** [reply]
Thanks for the reply! 
I tried to trick pybamm by increasing V_max a little bit, like from 4.2V to 4.2003V but still doesn't work... I think it is because of the dt_max, it said dt_max = 7.2 when it breaks and suggest a lower ones. But a low dt_max will prolong the computing time for dgradation modelling... Are there any other way like adaptoable dt_max setting?

**Reply by NLithium** [reply]
what about if you define ```initial_soc```?

**Reply by hamidbdr** [reply]
By applying 0.3C and the default period (60 seconds in your case) and starting from 4.2V the cell will overshoot the 4.2003V and break the experiment, about the dt_max (have a look at [max_step_decrease_counts)](https://pybamm.readthedocs.io/en/latest/source/solvers/casadi_solver.html) effect in the solver normally if the solver get inside the period and iterate over it may work, but as you said it will be time consuming. 

PS. I am new user of PyBaMM ;)

**Reply by RuiheLi** [reply]
> what about if you define `initial_soc`?

Thanks! I think that won't help as the cell will become fully-charged at the end anyway...

**Comment by valentinsulzer** [comment]
I think what's happening is that the simulation starts a bit above 4.2V, so it doesn't see the 4.2V limit. This is irrespective of the period or dt_max. We should hard-code this as a special case. I'll transfer this to an issue

**Reply by DrSOKane** [reply]
I don't see how it can start above 4.2 V. If there's a rest period after the CV hold, the voltage should fall, not rise.

What happens if you set inplace=True?

**Reply by valentinsulzer** [reply]
I mean instantaneously when a current is applied. If the OCV is at 4.2V, and you ask for 0.3C of current, there will be some overpotential that takes the voltage above 4.2V

**Reply by DrSOKane** [reply]
There is another possible cause, which is that Chen et al. didn't actually measure the open-circuit potential in the region around x=1 (full lithiation) and so PyBaMM assumes it to be completely flat in this region. The Chen2020_plating set gets around this problem by using the older lookup table, which is not flat in this region, at the expense of speed.

**Reply by RuiheLi** [reply]
Thanks so much for the helpful comments! I think Tino is right about the overpotential thing. 
I just tried first to cycle at 0.2C as followed: 
`import pybamm as pb;import matplotlib.pyplot as plt;import os;
Model_0 = pb.lithium_ion.DFN(
    options={
        "particle": "Fickian diffusion", 
        "SEI":"interstitial-diffusion limited", 
        "SEI film resistance":"distributed", 
        "SEI porosity change":"true",
        "lithium plating":"partially reversible"
    }
)
V_max = 4.2;        V_min = 2.5;
cs_Neg_Init = 28920; Diff_SEI = 1.700000000000002e-20;    R_SEI = 2E5;   Bulk_Sol_Con = 4541.0;

chemistry = pb.parameter_sets.Chen2020_coupled
Para_0 = pb.ParameterValues(chemistry=chemistry)
#Para_0["Ambient temperature [K]"] = pybamm.InputParameter("T")
Para_0.update({"Upper voltage cut-off [V]": 4.20})
Para_0.update({"Lower voltage cut-off [V]": 2.50})
Para_0.update({'SEI resistivity [Ohm.m]':R_SEI}) ;
Para_0.update({'Outer SEI solvent diffusivity [m2.s-1]':Diff_SEI});  
Para_0.update({'Bulk solvent concentration [mol.m-3]':Bulk_Sol_Con});
Para_0.update({'Initial concentration in negative electrode [mol.m-3]':cs_Neg_Init }) # 
Para_0.update({"Lithium plating kinetic rate constant [m.s-1]": 1E-18})
Para_0.update({"SEI-plating coupling constant [s-1]": 2E-6})
Experiment_Breakin= pb.Experiment( [    # at 25 degreeC
    (f"Discharge at 0.2 C until {V_min} V",
    "Rest for 4 hours",
    f"Charge at 0.2 C until {V_max} V" , 
    f"Hold at {V_max} V until C/100",
    "Rest for 2 hours")        ]   *5
    )  
Sim_0 = pb.Simulation(
    Model_0, experiment = Experiment_Breakin,
    parameter_values = Para_0,solver = pb.CasadiSolver(),)       
Sol_0 = Sim_0.solve();
t = Sol_0["Time [h]"].entries
Q = Sol_0["Discharge capacity [A.h]"].entries
V = Sol_0["Terminal voltage [V]"].entries
Sol_0.cycles[-1].steps[-1]["Terminal voltage [V]"].entries`


I notice that the voltage (during the 2hour rest peroid after 4.2V CV) hold decrease from 4.1994V to 4.1917V, decreasing by 8mV. Then I tried to charge the cell from this state for 114 s without voltage limit as follows:
`Experiment_0p3C_charge= pb.Experiment( [ (f"Charge at 0.3 C for 114 s (1 second period)", ) ] )
Experiment_0p1C_charge= pb.Experiment( [ (f"Charge at 0.1 C for 114 s (1 second period)", ) ] )
Experiment_0p01C_charge= pb.Experiment( [ (f"Charge at 0.01 C for 114 s (1 second period)", ) ] )

Model_1 = Model_0.set_initial_conditions_from(Sol_0)
sim1 = pb.Simulation(Model_1, parameter_values=Para_0, experiment=Experiment_0p3C_charge)
sol1 = sim1.solve()

Model_2 = Model_0.set_initial_conditions_from(Sol_0)
sim2 = pb.Simulation(Model_2, parameter_values=Para_0, experiment=Experiment_0p1C_charge)
sol2 = sim2.solve()

Model_3 = Model_0.set_initial_conditions_from(Sol_0)
sim3 = pb.Simulation(Model_3, parameter_values=Para_0, experiment=Experiment_0p01C_charge)
sol3 = sim3.solve()
t_1 = sol1["Time [h]"].entries
V_1 = sol1["Terminal voltage [V]"].entries
t_2 = sol2["Time [h]"].entries
V_2 = sol2["Terminal voltage [V]"].entries
t_3 = sol3["Time [h]"].entries
V_3 = sol3["Terminal voltage [V]"].entries

plt.plot(t_1, V_1,label='Charge at 0.3C')
plt.plot(t_2, V_2,label='Charge at 0.1C')
plt.plot(t_3, V_3,label='Charge at 0.01C')
plt.title('Charge after 100%SOC & 4.19174V') 
plt.legend(prop={'family':'Times New Roman','size':fs-3},loc='best',frameon=False)

print(V_1[0],V_2[0],V_3[0])`


The result is as follows:
![image](https://user-images.githubusercontent.com/84007676/165943967-4c46990d-7541-4690-8ac8-e8fc472506be.png)
It is now very clear that, at the last moment of the 2 hours rest, the voltage is 4.1917V, but at the first moment of the next charge step, the voltage is 4.1923V, 4.1974V, and 4.2088V for 0.01C, 0.1C, and 0.3C charge, respectively. Therefore, this can probably be ascribed to the issue of the solver.

I wonder whether we can set the solver so that:
1. under such conditions, the voltage starts extactly from the last moment, 
2. or the current can be automatically reduced to small enough,
3. or, PyBaMM can realize that this step cannot be operated and then skip this step.

I think this is a really common difficulty when we try to do GITT on aged cells when lots of steps will be skipped immediately.

**Reply by RuiheLi** [reply]
> 

Regarding the OCP, I just checked the voltage I used. It uses the loop-up table with linear extrapolation which is extactly the one used in Chen2020_plating.
